### ISSUE TITLE
how to pass output parameter to ms-sql stored procedure using knex.

### ISSUE THREAD

**Issue by so-abahadur** [tgriesser/knex#2764]
# Environment

Knex version: 0.15.0
Database + version: MS-SQL
OS: Windows

# Feature discussion / request

1. Explain what is your use case
we are using Knex in node js to call MS-SQL raw query and stored procedure. In MS-SQL, we can pass both IN and OUTPUT parameter in stored procedure. Now, i am stuck, how to pass output parameter to ms-sql stored procedure.
Procedure parameter is like this :
GetEmployeeCount (@totalEmployee BIT OUTPUT);

2. Explain what kind of feature would support this

3. Give some API proposal, how the feature should work

# Missing / erroneus documentation

Send issue to documentation repo, or fix it and send PR https://github.com/knex/documentation

# Questions about how to use knex

Github issues are for knex development. Please send questions how to use knex to
stackoverflow or ask about it in gitter chat.

**Comment by elhigu** [comment]
How does mssql driver support that? For oracle there was certain object syntax that must be passed to bindings to tell the driver that it is an output parameter.

**Comment by so-abahadur** [comment]
like this =>
const knex = knexInstance({
    client: 'mssql',
    connection: {
        host: environment_1.Database.host,
        user: environment_1.Database.username,
        password: environment_1.Database.password,
        database: environment_1.Database.dbName,
        options: {
            encrypt: true
        }
    }
});

**Comment by Janso123** [comment]
tediousjs/node-mssql
```
var request = new sql.Request(connection);
request.input('id', sql.Int, 5);
request.output('result', sql.VarChar(50));
request.execute('SP_TEST3', function(err, recordsets, returnValue) {
     console.log(request.parameters.result.value);
});
```

**Comment by thiagotigaz** [comment]
Any update here? I'm having the same issue, I need to read the returned parameters

**Comment by pate313373** [comment]
I agree, calling stored procedures with both input and output parameters would be quite a basic feature needed.

**Comment by elhigu** [comment]
@pate313373 would you like provide PR for this needed basic feature.

**Comment by pate313373** [comment]
Basically, the example [at tedious](https://github.com/tediousjs/tedious/blob/master/examples/stored-procedure-with-parameters.js) should be more than sufficient to implement a `StoredProcedure `class for knex.

Unfortunately, I have 3 projects with thight deadlines at the moment, so I'm using knex and the npm package mssql natively at the same time. - However, I really don't like this workaround.

Would you consider it, if it supports MSSQL only? Because if I add this feature, most likely only for MSSQL.

**Comment by kibertoad** [comment]
@pate313373 As long as API is DB-agnostic and could potentially be implemented for other databases (which should throw an error for now), that would be an option.

**Comment by NamedArray** [comment]
@pate313373 How do you add the example you'd mentioned above as a class to an existing knex instance?

**Comment by pate313373** [comment]
> 
> 
> @pate313373 How do you add the example you'd mentioned above as a class to an existing knex instance?

I did not use the existing knex instance, I rather connected to the DB in parallel with tedious.

**Comment by mercmobily** [comment]
Knex does not currently provide a first-class API for output parameters with MSSQL, but it can be supported either by passing through to the driver or by adding a small helper in the MSSQL dialect.

The `tedious/mssql` drivers support named input/output binds at the driver level (e.g. `request.output('result', sql.VarChar(50))`). Knex’s `raw()` passes SQL and bindings to the driver, but there is no Knex-level `StoredProcedure` helper to make output parameters ergonomic across dialects. Two viable paths:
- Short term: call the driver directly for complex stored procedure calls, alongside Knex. This is what several users in the thread have done.
- Medium term: add a DB-agnostic stored procedure invocation helper that throws on unsupported dialects. For MSSQL, it would translate to tedious’s `Request.input` and `Request.output` methods under the hood. For others... oh my, I did some research, and it's possible, but it's definitely a lot of work. But, in my opinion, it would definitely be worth it! 

This is the kind of addition we can get to once we have the smaller issues under control.
